backend/user-service/src/controllers/authController.ts

```diff
@@ -375,3 +400,83 @@ export async function updateAccessToken(req: Request, res: Response) {
     }
   }
 }
+
+export const updateUserProfile: RequestHandler[] = [
+  body("username").notEmpty().withMessage("Username cannot be empty"),
+  body("email")
+    .notEmpty()
+    .withMessage("Email cannot be empty!")
+    .isEmail()
+    .withMessage("Invalid email format!"),
+  async (req, res) => {
+    if (!validationResult(req).isEmpty()) {
+      res.status(400).json({ errors: validationResult(req).array() });
+      return;
+    }
+
+    try {
+      const user = req.user;
+
+      if (!user) {
+        return res.status(401).json({ message: "User not authenticated" });
+      }
+
+      const { username, email, languages } = req.body;
+
+      const languageIds = [];
+
+      for (const language of languages) {
+        let existingLanguages = await prisma.language.findMany({
+          where: {
+            language: language.language,
+          },
+        });
+
+        const existingLanguage = existingLanguages[0];
+
+        if (existingLanguage) {
+          languageIds.push(existingLanguage.id);
+        } else {
+          const newLanguage = await prisma.language.create({
+            data: { language: language },
+          });
+          languageIds.push(newLanguage.id);
+        }
+      }
+
+      const updatedUser = await prisma.user.update({
+        where: { id: user.id },
+        include: {
+          languages: true,
+        },
+        data: {
+          username: username,
+          email: email,
+          languages: {
+            set: [],
+            connect: languageIds.map((id) => ({ id: id })),
+          },
+        },
+      });
```

Comment:
nit: hashed password getting returned to user
<img width="980" alt="image" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/34371483/5177dc10-5082-43f7-acbc-043ffb53b15e">

Reply:
wait my bad sorry i think i was running something outdated